src/uu/expr/src/syntax_tree.rs

```diff
@@ -110,67 +105,66 @@ impl NumericOp {
     }
 }
 
 impl StringOp {
-    fn eval(&self, left: &AstNode, right: &AstNode) -> ExprResult<String> {
+    fn eval(&self, left: &AstNode, right: &AstNode) -> ExprResult<NumOrStr> {
         match self {
             Self::Or => {
                 let left = left.eval()?;
-                if is_truthy(&left) {
+                if is_truthy(&left.to_string()) {
                     return Ok(left);
                 }
                 let right = right.eval()?;
-                if is_truthy(&right) {
+                if is_truthy(&right.to_string()) {
                     return Ok(right);
                 }
-                Ok("0".into())
+                Ok(NumOrStr::Num(BigInt::from(0)))
             }
             Self::And => {
                 let left = left.eval()?;
-                if !is_truthy(&left) {
-                    return Ok("0".into());
+                if !is_truthy(&left.to_string()) {
+                    return Ok(NumOrStr::Num(BigInt::from(0)));
                 }
                 let right = right.eval()?;
-                if !is_truthy(&right) {
-                    return Ok("0".into());
+                if !is_truthy(&right.to_string()) {
+                    return Ok(NumOrStr::Num(BigInt::from(0)));
                 }
                 Ok(left)
             }
             Self::Match => {
                 let left = left.eval()?;
                 let right = right.eval()?;
-                let re_string = format!("^{}", &right);
+                let re_string = format!("^{}", right.to_string());
                 let re = Regex::with_options(
                     &re_string,
                     RegexOptions::REGEX_OPTION_NONE,
                     Syntax::grep(),
                 )
                 .map_err(|_| ExprError::InvalidRegexExpression)?;
-                Ok(if re.captures_len() > 0 {
-                    re.captures(&left)
+                Ok(NumOrStr::Str(if re.captures_len() > 0 {
+                    re.captures(&left.to_string())
                         .map(|captures| captures.at(1).unwrap())
                         .unwrap_or("")
                         .to_string()
                 } else {
-                    re.find(&left)
+                    re.find(&left.to_string())
                         .map_or("0".to_string(), |(start, end)| (end - start).to_string())
-                })
+                }))
             }
             Self::Index => {
                 let left = left.eval()?;
                 let right = right.eval()?;
-                for (current_idx, ch_h) in left.chars().enumerate() {
-                    for ch_n in right.chars() {
+                for (current_idx, ch_h) in left.to_string().chars().enumerate() {
+                    for ch_n in right.to_string().chars() {
                         if ch_n == ch_h {
-                            return Ok((current_idx + 1).to_string());
+                            return Ok(NumOrStr::Num(BigInt::from(current_idx + 1)));
```

Comment:
It might be nice to implement `From<usize>`, `From<BigInt>` and `From<String>` for `NumOrStr`